substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+	///
+	/// Epoch X first N-th ticket has key (X mod 2, N)
+	///
+	/// Note that the ticket's index doesn't directly correspond to the slot index within the epoch.
+	/// The assigment is computed dynamically using an *outside-in* strategy.
+	///
+	/// Be aware that entries within this map are never removed, only overwritten.
+	/// Last element index should be fetched from the [`TicketsMeta`] value.
+	#[pallet::storage]
+	pub type TicketsIds<T> = StorageMap<_, Identity, (u8, u32), TicketId>;
+
+	/// Tickets to be used for current and next epoch.
+	#[pallet::storage]
+	pub type TicketsData<T> = StorageMap<_, Identity, TicketId, TicketBody>;
+
+	/// Next epoch tickets unsorted segments.
+	///
+	/// Contains lists of tickets where each list represents a batch of tickets
+	/// received via the `submit_tickets` extrinsic.
+	///
+	/// Each segment has max length [`SEGMENT_MAX_SIZE`].
+	#[pallet::storage]
+	pub type UnsortedSegments<T: Config> =
+		StorageMap<_, Identity, u32, BoundedVec<TicketId, ConstU32<SEGMENT_MAX_SIZE>>, ValueQuery>;
+
+	/// The most recently set of tickets which are candidates to become the next
+	/// epoch tickets.
+	#[pallet::storage]
+	pub type SortedCandidates<T> =
+		StorageValue<_, BoundedVec<TicketId, MaxTicketsFor<T>>, ValueQuery>;
+
+	/// Parameters used to construct the epoch's ring verifier.
+	///
+	/// In practice: Updatable Universal Reference String and the seed.
+	#[pallet::storage]
+	#[pallet::getter(fn ring_context)]
+	pub type RingContext<T: Config> = StorageValue<_, vrf::RingContext>;
+
+	/// Ring verifier data for the current epoch.
+	#[pallet::storage]
+	pub type RingVerifierData<T: Config> = StorageValue<_, vrf::RingVerifierData>;
+
+	/// Slot claim vrf-preoutput used to generate per-slot randomness.
+	///
+	/// The value is ephemeral and is cleared on block finalization.
+	#[pallet::storage]
+	pub(crate) type ClaimTemporaryData<T> = StorageValue<_, vrf::VrfOutput>;
+
+	/// Genesis configuration for Sassafras protocol.
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		/// Genesis authorities.
+		pub authorities: Vec<AuthorityId>,
+		/// Genesis epoch configuration.
+		pub epoch_config: EpochConfiguration,
+		/// Phantom config
+		#[serde(skip)]
+		pub _phantom: sp_std::marker::PhantomData<T>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			EpochConfig::<T>::put(self.epoch_config);
+			Pallet::<T>::initialize_genesis_authorities(&self.authorities);
+
+			#[cfg(feature = "construct-dummy-ring-context")]
+			{
+				debug!(target: LOG_TARGET, "Constructing dummy ring context");
+				let ring_ctx = vrf::RingContext::new_testing();
+				RingContext::<T>::put(ring_ctx);
+				Pallet::<T>::update_ring_verifier(&self.authorities);
+			}
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(block_num: BlockNumberFor<T>) -> Weight {
```

Comment:
I guess we'll need similar logic as BABE to track initialization once we add the integration with session (i.e. implement `ShouldEndSession` trait).

Reply:
Yeah. I already have it in the main PR. But I provided this PR without thinking (too much) about session integration and other additions to keep it as much independent as possible